keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,987 @@
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP aims to add a `.status.schedulePaused` field to Pod spec, to mark a Pod's schedule readiness.
+Integrators can mutate this field to instruct scheduler whether it's ready for scheduling or not.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Pods are naturally perceived ready for scheduling once created. Kubernetes scheduler does
+its due diligence to find nodes to place all pending Pods. However, in a real-world case,
+some Pods may stay in a "miss-essential-resources" state for a long period. These pods,
+actually churn the scheduler (and downstream integrators like Cluster AutoScaler) in an
+unnecessary manner.
+
+Lacking the knob to flag Pods as schedule-ready/unready wastes scheduling cycles on retrying
+Pods that are determined to be unschedulable. As a result, it delays the scheduling of other Pods,
+and also lowers the overall scheduling throughput. Moreover, it imposes restrictions to vendors to
+develop some in-house features (such as hierarchical quota) natively.
+
+This proposal comes up with an extensible mechanics to ensure Pods are scheduled only when they are
+semantically ready.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define API to mark Pod as schedule-paused/unpaused.
+- Design a new Enqueue extension point to customize Pod's queueing behavior.
+- A default enqueue plugin to honor the new API semantics.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Update the Pod's condition if it carries `.status.schedulePaused=true`.
+- Focus on in-house use-cases of the Enqueue extension point.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose a new field `.status.schedulePaused` to the Pod API. The field is defaulted to `false`.
```

Comment:
Any thoughts around an alternative way to specify the desire in the `podSpec` (potentially through a dedicated sub-resource) rather than in `podStatus`? Was thinking of something along the lines of https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1287-in-place-update-pod-resources/README.md#api-changes.

Reply:
I was expecting this to be in spec as well.  This field is mutated prior to scheduling, and in general kubelet won’t modify spec.  It’s very similar to schedulerName and paused for me, which are both spec fields.

Would we ever parameterize or modify this value?

Did you consider schedulingPaused over schedulePaused?  Do we have any other examples of `<x><State>` naming we can look to?

Reply:
I chose status was b/c `pod/status` is an endpoint that would cause less conflict than `pod` itself. However, Clayton pointed out in a comment below users may set this flag in PodTemplate. In that case, `podStauts` doesn't fit anymore. I'm going to update the design.

(btw: it's a different story with preemption which is wired by `.status.nominatedNodeName` as that's always a Pod-level operation)

Regarding associating it with a sub-resource, I assume it's wired to a particular pod? like `/pods/<name>/pause`. If so, it may be of less value as we cannot update a batch of Pods' pause flag via PodTemplate. (correct me if I'm wrong).

> Would we ever parameterize or modify this value?

The value is only used when the pod is pending (.spec.nodeName=nil).

> Did you consider schedulingPaused over schedulePaused?

No strong option on the naming. `schedulingPaused` is fine with me as well.

Reply:
Good point about being able to set it through PodTemplates for controllers.

Reply:
Regarding parameterization, I should have been clearer:

> Will we ever have multiple possible values, or is this a true boolean?

Precedent for `paused` is true boolean, just want to be absolutely sure you don't imagine different ways of handling the pause (multiple phases, conditionals) such that this should be a) an enum or b) a struct with deeper details.

Let me think on the naming a bit more and ask for feedback from API review - I like `<verb>Paused` a bit more, but I'll want some opinions.

Reply:
It's a boolean, and we will prevent the flag from being flipped back and forth.